pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+    """
+
+    if not isinstance(wires, Iterable):
+        raise ValueError("Wires must be passed as a list of integers; got {}.".format(wires))
+
+    if not len(basis_state) == len(wires):
+        raise ValueError(
+            "Number of qubits must be equal to the number of wires, which is {}; "
+            "got {}.".format(len(wires), len(basis_state))
+        )
+
+    if any([x not in [0, 1] for x in basis_state]):
+        raise ValueError("Basis state must only consist of 0s and 1s, got {}".format(basis_state))
+
+    for wire, state in zip(wires, basis_state):
+        if state == 1:
+            qml.PauliX(wire)
+
+
+def _matrix_M_entry(row, col):
+    """Returns one entry for the matrix that maps alpha to theta.
+
+    Args:
+        row (int): one-based row number
+        col (int): one-based column number
+
+    Returns:
+        (float): transformation matrix entry at given row and column
+    """
+    # (col >> 1) ^ col is the Gray code of col
+    b_and_g = row & ((col >> 1) ^ col)
+    sum_of_ones = 0
+    while b_and_g > 0:
+        if b_and_g & 0b1:
+            sum_of_ones += 1
+
+        b_and_g = b_and_g >> 1
+
+    return (-1) ** sum_of_ones
+
+
+def _compute_theta(alpha):
+    """Calculates the rotation angles from the alpha vector.
+
+    Args:
+        alpha (array[float]): alpha parameters
+
+    Returns:
+        (array[float]): rotation angles theta
+    """
+    k = np.log2(alpha.shape[0])
+    factor = 2 ** (-k)
+
+    theta = sparse.dok_matrix(alpha.shape, dtype=np.float64)  # type: sparse.dok_matrix
+
+    for row in range(alpha.shape[0]):
+        # Use transpose of M:
+        entry = sum([_matrix_M_entry(col, row) * a for (col, _), a in alpha.items()])
+        entry *= factor
+        if abs(entry) > 1e-6:
+            theta[row, 0] = entry
+
+    return theta
+
+
+def _uniform_rotation_dg(gate, alpha, control_wires, target_wire):
+    """Applies a given inverse rotation to the target qubit
+    that is uniformly controlled by the control qubits.
+
+    Args:
+        gate (qml.Operation): gate to be applied, needs to have exactly
+            one parameter
+        alpha (array[float]): alpha parameters
+        control_wires (array[int]): wires that act as control
+        target_wire (int): wire that acts as target
+    """
+
+    theta = _compute_theta(alpha)  # type: sparse.dok_matrix
+
+    gray_code_rank = len(control_wires)
+
+    if gray_code_rank == 0:
+        gate(theta[0, 0], wires=[target_wire])
+        return
+
+    code = gray_code(gray_code_rank)
+    num_selections = len(code)
+
+    control_indices = [
+        int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % num_selections], 2)))
+        for i in range(num_selections)
+    ]
+
+    for i, control_index in enumerate(control_indices):
+        gate(theta[i, 0], wires=[target_wire])
+        qml.CNOT(wires=[control_wires[control_index], target_wire])
+
+
+def _unirz_dg(alpha, control_wires, target_wire):
```

Comment:
Elaborating/splitting the `unirz` part of the function name such as `uni_rz`, `uni_zrot` or `uniformly_controlled_zrot` could also be nice for the reader